core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java

```diff
@@ -0,0 +1,298 @@
+
+  /** Used to store latest catch-up thread. */
+  private AbstractCatchupThread mCatchupThread;
+
+  /** Used to synchronize buffer state. */
+  private ReentrantLock mStateLock = new ReentrantLock(true);
+
+  /**
+   * Creates a buffered applier over given journals.
+   *
+   * @param journals journals
+   * @param journalSinks journal sinks
+   */
+  public BufferedJournalApplier(Map<String, RaftJournal> journals,
+      Supplier<Set<JournalSink>> journalSinks) {
+    mJournals = journals;
+    mJournalSinks = journalSinks;
+  }
+
+  /**
+   * Processes given journal entry for applying. Entry could be applied or buffered based on
+   * buffer's state.
+   *
+   * @param journalEntry the journal entry
+   */
+  public void processJournalEntry(Journal.JournalEntry journalEntry) {
+    try (LockResource stateLock = new LockResource(mStateLock)) {
+      if (mSuspended) {
+        // New entry submissions are monitored by catch-up threads.
+        synchronized (mSuspendBuffer) {
+          mSuspendBuffer.offer(journalEntry);
+          mSuspendBuffer.notifyAll();
+        }
+      } else {
+        applyToMaster(journalEntry);
+      }
+    }
+  }
+
+  /**
+   * @return {@code true} if this applier was suspended
+   */
+  public boolean isSuspended() {
+    try (LockResource stateLock = new LockResource(mStateLock)) {
+      return mSuspended;
+    }
+  }
+
+  /**
+   * Suspend the applier.
+   *
+   * After this call, journal entries will be buffered until {@link #resume()} or
+   * {@link #catchup(long)} is called.
+   *
+   * @throws IOException
+   */
+  public void suspend() throws IOException {
+    try (LockResource stateLock = new LockResource(mStateLock)) {
+      Preconditions.checkState(!mSuspended, "Already suspended");
+      LOG.info("Suspending state machine.");
+      mSuspended = true;
+      LOG.info("Suspended state machine at sequence: {}", mLastAppliedSequence);
+    }
+  }
+
+  /**
+   * Resumes the applier. This method will apply all buffered entries before returning.
+   *
+   * @throws IOException
+   */
+  public void resume() throws IOException {
+    try (LockResource stateLock = new LockResource(mStateLock)) {
+      Preconditions.checkState(mSuspended, "Not suspended");
+      Preconditions.checkState(!mResumeInProgress, "Resume in progress");
+      mResumeInProgress = true;
+      LOG.info("Resuming state machine.");
+    }
+
+    // Cancel catching up thread if active.
+    if (mCatchupThread != null && mCatchupThread.isAlive()) {
+      mCatchupThread.cancel();
+      mCatchupThread.waitTermination();
+    }
+
+    /**
+     * Applies all buffered entries.
+     *
+     * It doesn't block state until
+     *   -> buffer contains few elements ( RESUME_LOCK_BUFFER_SIZE_WATERMARK )
+     *   -> was running for a long time  ( RESUME_LOCK_TIME_LIMIT_MS         )
+     */
+    try {
+      // Mark resume start time.
+      Instant resumeStartTime = Instant.now();
```

Comment:
Users have brought up performance issues with `Instant.now`, can we use `System.nanoTime`?